Mono.Nat/Pmp/Searchers/PmpSearcher.cs

```diff
@@ -109,24 +111,28 @@ internal static PmpSearcher Create ()
 
         }
 
-        protected override async Task SearchAsync (IPAddress gatewayAddress, TimeSpan? repeatInterval, CancellationToken token)
+        protected override async void SearchAsync (IPAddress gatewayAddress, TimeSpan? repeatInterval, CancellationToken token)
         {
-            if (token == CancellationToken.None) {
-                token = Cancellation.Token;
-            }
-
             do {
-                var currentSearch = CancellationTokenSource.CreateLinkedTokenSource (token);
-                Interlocked.Exchange (ref CurrentSearchCancellation, currentSearch)?.Cancel ();
+                CurrentSearchCancellation?.Cancel ();
+                var currentSearch = CurrentSearchCancellation = CancellationTokenSource.CreateLinkedTokenSource (token);
```

Comment:
This is definitely not thread-safe though.If it's called from different threads, this can race in certain conditions. Prefer using `Interlocked.CompareExchange` here and then you got a sure-fire way to cancel _and_ dispose the current token source.

Reply:
That kind of thread safety isn't a concern in this specific place!

These classes are internal so there's no direct access to them. The way to access them is to call methods on the static class, `NatUtility`, and these methods are wrapped in a lock. 

As such, we're guaranteed this part of the code will always be singlethreaded. That said, thread safety is a concern elsewhere :)